src/model/Document.cpp

```diff
@@ -407,13 +407,6 @@ auto Document::operator=(const Document& doc) -> Document& {
     buildContentsModel();
     updateIndexPageNumbers();
 
-    bool lastLock = tryLock();
-    unlock();
-    this->handler->fireDocumentChanged(DOCUMENT_CHANGE_COMPLETE);
```

Comment:
I don't understand why we should fire document changed here. It's already called in Control::newFile. And there we for sure know that we have the lock.

Reply:
Agreed. It's also not the model's role the lock itself. The caller should lock the document before calling the operator.
Please make sure that this is the case anywhere the operator is used.

Reply:
> It's also not the model's role the lock itself

It's actually the other way round. Optimally, lockable's should not be accessible from the public.

Reply:
I agree on that. Document should probably be implemented as monitor. I guess this is what @LittleHuba meant:
> We could actually overload the -> operator for the wrapper class. That is a common way to do something like that.

The only problem I see with that is that we might get some overhead if a lot of functions are called on document in a series. Then it would be better to lock the document, do our stuff and then return the lock. We could fix that with a reentrant lock, but this would somewhat break the idea of monitors.

Reply:
The problem is actually that you can't do multiple actions atomically. A better solution would be a "Monitor" that resembles a RAII-lock. It should enable exclusive access to the document during its lifetime and relinquish the lock automatically during destruction. We should make it reentrant to cover the current use-cases.